user/tests/wiring/no_fixture/acm.cpp

```diff
@@ -0,0 +1,155 @@
+    WiFi.prefer();
+    WiFi.connect();
+    assertTrue(waitFor(WiFi.ready, WIFI_CONNECT_TIMEOUT_MS));
+#endif // HAL_PLATFORM_WIFI && !HAL_PLATFORM_WIFI_SCAN_ONLY
+
+#if HAL_PLATFORM_CELLULAR
+    expectedNetwork = Cellular;
+    Cellular.prefer();
+    Cellular.connect();
+    assertTrue(waitFor(Cellular.ready, HAL_PLATFORM_CELLULAR_CONN_TIMEOUT));
+#endif // HAL_PLATFORM_CELLULAR
+
+    Particle.connect();
+    assertTrue(waitFor(Particle.connected, HAL_PLATFORM_MAX_CLOUD_CONNECT_TIME));
+    assertEqual(Particle.connectionInterface(), expectedNetwork);
+}
+
+
+#if PLATFORM_ID == PLATFORM_MSOM
+test(ACM_03_prefer_one_network_only) {
+	WiFi.prefer();
+	assertFalse(Cellular.isPreferred());
+	Cellular.prefer();
+	assertFalse(WiFi.isPreferred());
+}
+
+test(ACM_04_prefer_moves_cloud_connection_when_set) {
+    // Default preference
+	Network.prefer();
+	
+    // Connect all interfaces
+    Network.connect();
+    Particle.connect();
+    assertTrue(waitFor(Particle.connected, HAL_PLATFORM_MAX_CLOUD_CONNECT_TIME));
+
+    auto connectedNetwork = Particle.connectionInterface();
+
+    // Prefer whatever network we did not connect on in order to force the cloud connection to switch
+    if (connectedNetwork == Cellular) {
+	    assertTrue(waitFor(WiFi.ready, WIFI_CONNECT_TIMEOUT_MS));
+    	WiFi.prefer();
+    } else {
+    	assertTrue(waitFor(Cellular.ready, HAL_PLATFORM_CELLULAR_CONN_TIMEOUT));
+    	Cellular.prefer();
+    }
+
+    // Verify that we disconnect and reconnect on the newly preferred network
+    assertTrue(waitForNot(Particle.connected, DISCONNECT_TIMEOUT_MS));
+    assertTrue(waitFor(Particle.connected, HAL_PLATFORM_MAX_CLOUD_CONNECT_TIME));
+    assertNotEqual(connectedNetwork, Particle.connectionInterface());
+}
+
+test(ACM_05_cloud_connection_moves_to_preferred_network_when_available) {
+    // Disconnect WiFi
+    WiFi.disconnect();
+    Particle.disconnect();
+    assertTrue(waitForNot(Particle.connected, DISCONNECT_TIMEOUT_MS));
+    assertTrue(waitForNot(WiFi.ready, DISCONNECT_TIMEOUT_MS));
+
+    // Connect Cellular
+	Cellular.connect();
+	Particle.connect();
+    assertTrue(waitFor(Cellular.ready, HAL_PLATFORM_CELLULAR_CONN_TIMEOUT));
+	assertTrue(waitFor(Particle.connected, HAL_PLATFORM_MAX_CLOUD_CONNECT_TIME));
+
+    // Verify we are using celluar
+	auto connectedNetwork = Particle.connectionInterface();
+	assertEqual(Cellular, Particle.connectionInterface());
+
+    // Prefer wifi, switch to it
+	WiFi.prefer();
+	WiFi.connect();
+
+    // We should see the particle socket disconnect and reconnect
+	assertTrue(waitFor(WiFi.ready, WIFI_CONNECT_TIMEOUT_MS));
+	assertTrue(waitFor(Particle.disconnected, HAL_PLATFORM_MAX_CLOUD_CONNECT_TIME));
+	assertTrue(waitFor(Particle.connected, HAL_PLATFORM_MAX_CLOUD_CONNECT_TIME));
+
+    // Verify we are on wifi now
+	assertEqual(WiFi, Particle.connectionInterface());
+} 
+
+test(ACM_06_cloud_connection_fails_over_automatically) {
+    Particle.disconnect();
```

Comment:
I don't think I'm understanding this test...
I feel like I want to add waitForNot Particle.connected here, but it seems something tricky is being done to set the state of everything first... then wait for Particle.connected() which most likely will be connect still.
It was previously connected on WiFi, so it will disconnect that... and then in the background it tries to connect to Cellular, and we test that it goes through the disconnect/connect sequence and that it's not connected to WiFi anymore.  I feel like this one can be tightened up to more clearly show the failover response... maybe even force it to be on WiFi first and switch to Cellular, and vice versa.

Reply:
Yeah this test is wrong, hopefully clearer now